### ISSUE TITLE
I have a problem (may be a misunderstanding) with type inference for Array 

### ISSUE THREAD

**Issue by g3ortega** [crystal-lang/crystal#2758]
Hi. let's say I have this Array of Arrays

```crystal
  training = [
    [36.6, "healthy"],
    [37.1, "sick"],
    [38.3, "sick"],
    [36.2, "healthy"],
    [40.2, "sick"],
    [50.1, "really sick"],
  ]
```

And this class:

```crystal
  class ID3Tree

    @attributes : Array(String)
    @data : Array(Array(Float64 | String) | Array(Int32 | String))
    @default : String
    @type : Symbol
  
    def initialize(@attributes, @data, @default, @type)
      @used = {} of String => Array(Int32)
      @tree = {} of String => Array(Int32)
    end
    
  end
```

As you can see, data is declared as a kind of nested Array, but in the inner one you can pass it an Array containing a float or a string, or, an Array containing an integer or string. I did that because sometimes the training data set (that's the @data attribute) sometimes include integers and other times floats.

When I run the program with a dataset containing both kind of arrays (with integer and float variations), I have no problems. But when I include only one kind I receive an error like this:

```
in ./src/decisiontree/id3_tree.cr:17: instance variable '@data' of DecisionTree::ID3Tree must be Array(Array(Float64 | String) | Array(Int32 | String)), not Array(Array(Float64 | String))
```

It's interesting, because that would be a good error message when I pass more than one kind of array, and in any case, I declared the right type in the instance variable declaration. If I'm not wrong and you want to reproduce the issue, you can check [this repo](https://github.com/g3ortega/test_issue).

Regards.

**Comment by asterite** [comment]
This is not a bug, read #1297 to understand why it doesn't work.

Basically:

```crystal
@x = Array(Array(Float64 | String) | Array(Int32 | String))
@y = Array(Array(Float64 | String)).new

# Now we make @x have the contents of @y
@x = @y

# Valid, because @x is Array(Array(Float64 | String) | Array(Int32 | String))
@x << Array(Int32 | String).new

# But... we just inserted an Array(Int32 | String) into @y, which doesn't allow that type,
# so we basically bypassed the type system
``



Basically, your variable is of type `Array(Array(Float64 | String) | Array(Int32 | String))`. Imagine you assign a variable of type `Array(Array(Float64 | String))` to it, but keep a reference to it. Then

**Comment by g3ortega** [comment]
Thank you @asterite, I'm new with Crystal, and I'm learning a lot. I think that deserve a section in documentation, when I understand it better I could help with that.

**Comment by asterite** [comment]
@g3ortega As a side note, if you need such "too many unions" type it might mean the problem can be modelled in a simpler way. Do you want to explain why you need that?

Explaining that in the docs is probably good, though right now it's not clear how we'll solve it. Maybe we'll introduce explicit covariance/contravariance annotations.
